[Lim Ngian Xin Terry] iP

TypeDefinition opened the pull request:
# Delphine (Duke, but cuter)
> "[uwu](https://www.youtube.com/watch?v=7mBqm8uO4Cg)" - Some Anime Girl

Delphine frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- it is ~~GOOD~~ ~~GOODER~~ GOODER-EST! 

All you need to do is,
1. download it from here.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!!!**

Features:
- [x] Cute anime girl mascot :heart_eyes: 
- [x] Managing tasks
- [ ] Voice over (coming soon) :speaking_head: 

___

If you want to extend it, it's super easy! Here is how you add more commands:

```java
public ChatBot() {
    this.parser = new Parser();
    this.parser.addCommand(new TerminateCommand());
    this.parser.addCommand(new DeadlineCommand());
    this.parser.addCommand(new DeleteCommand());
    this.parser.addCommand(new EventCommand());
    this.parser.addCommand(new ListCommand());
    this.parser.addCommand(new MarkCommand());
    this.parser.addCommand(new ToDoCommand());
    this.parser.addCommand(new UnmarkCommand());
    this.parser.addCommand(new ResetCommand());
    this.parser.addCommand(new FindCommand());
}
```

Commit:
Finished Level-1.

Commit:
Finished Level-2.

Commit:
Finished Level-3 & Level-4.

Commit:
Completed A-TextUiTesting & Level-5.

Commit:
Finished Level-6.

Commit:
Fixed a bug where the program will crash if a blank input is given.

Commit:
Change boolean variable name to fit Java coding standard.

Commit:
Remove unecessary function

Commit:
Add saving and rename chat bot to Delphine.

Commit:
Add date & time support for Event & Deadline.

Commit:
Merge remote-tracking branch 'origin/branch-Level-8'

Commit:
Finish A-MoreOOP

Commit:
Finish A-Packages

The project has already been organized into packages since the
beginning.

Commit:
Add JUnit tests

Commit:
Fix compile error.

Commit:
Support building as JAR file.

The current method of accessing resource files only worked when running
the program from IntelliJ.
The program was not able to access resource files when ran as a JAR.
This caused the program to not be able to load, save or play audio.

Wrap all file streams with a buffered stream.

This lets the program to access resource files regardless of how it is
packaged.

Commit:
Finish Level-9

Commit:
Comply to coding iP standards

Commit:
Add JavaDoc

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge A-CodingStandard, A-JavaDoc and A-Level-9 and solve merge conflicts

Commit:
Modify find command to ignore casing

honganhcs commented on src/main/java/chatbot/ChatBot.java:
```diff
@@ -0,0 +1,88 @@
+    private static final String SAVE_FILE_PATH = "./data/save_file";
+    private static final String QUIT_TRIGGER = "bye";
+    private static final String LOGO =
+              "⣿⡟⠙⠛⠋⠩⠭⣉⡛⢛⠫⠭⠄⠒⠄⠄⠄⠈⠉⠛⢿⣿⣿⣿⣿⣿⣿⣿⣿⣿\n"
```
I like your design, good effort 😃

honganhcs commented on src/main/java/chatbot/command/CommandOutput.java:
```diff
@@ -0,0 +1,20 @@
+ * Output of commands.
+ */
+public class CommandOutput {
+    public final String output;
```
I noticed that in your command-related classes, your class variables are declared as public, which [violates information hiding](https://se-education.org/guides/conventions/java/intermediate.html#variables). Is it possible to declare them as private?
However, I like that you declared them as final so that they cannot be modified by other classes.

TypeDefinition replied:
As per the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#variables), the variables are declared as public since this is a _data class with no behaviour_.

honganhcs commented on src/main/java/chatbot/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+            desc = String.join(" ", Arrays.asList(input).subList(1, i));
+            date = LocalDate.parse(input[i + 1]);
+            time = LocalTime.parse(input[i + 2]);
+        } catch (Exception ignored) {
```
Maybe you could add a comment to explain why the exception is ignored? 😄

honganhcs commented on src/main/java/chatbot/util/Storage.java:
```diff
@@ -0,0 +1,63 @@
+     * @param <T>      the type of the serializable item
+     * @return returns true if the item was successfully saved, else returns false
+     */
+    public static <T extends Serializable> boolean Save(String filePath, T item) {
```
Minor point, but I think it would better to name these two functions as `save` and `load` (camelCase), unless you have a reason for capitalizing them 😆

TypeDefinition replied:
Thanks for pointing that out. I just simply missed them out.

honganhcs reviewed (commented):
Hi, I think your PR is looking good! Your implementation is clear, and you followed the coding standard very well. There are a few parts that I think can be changed to better follow the guidelines; otherwise, good job 😄😄

robinrojh commented on src/main/java/chatbot/task/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+     * @param saveFile the save file location
+     */
+    public static TaskList create(String saveFile) {
+        TaskList taskList = Storage.Load(saveFile);
```
Method names should be in camelCase.

robinrojh commented on src/main/java/chatbot/task/TaskList.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public Task remove(int index) {
+        Task ret = super.remove(index);
+        Storage.Save(saveFile, this);
```
Similarly, method names should be in camelCase.

robinrojh commented on src/main/java/chatbot/util/Ui.java:
```diff
@@ -0,0 +1,29 @@
+
+import javax.sound.sampled.AudioSystem;
+import javax.sound.sampled.Clip;
+import java.io.BufferedInputStream;
```
Imports for different packages can be separated with a newline.

robinrojh commented on src/main/java/chatbot/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,49 @@
+            while (i < input.length && !input[i].equals("/by")) {
+                ++i;
+            }
+            desc = String.join(" ", Arrays.asList(input).subList(1, i));
```
Could look a bit less complicated-looking if you separate the `Arrays.asList(input).subList(1, i)` part to another variable.

Commit:
Rename Storage.save & Storage.save function to use camelCasing.

Commit:
Add comments to explain why certain exceptions are ignored.

Commit:
Refactor "bye" into a command of its own. Parser now actually parses the input.

Commit:
Update README

Commit:
Add build.gradle

Commit:
Add gradle support

Commit:
Suppress warnings for statements already in a try-catch

Commit:
Add checkstyle

Commit:
Make build.gradle code style consistent.

Commit:
Add JavaFX dependencies top build.gradle

Commit:
Add basic GUI

Commit:
Split MainWindow into MVC format

Commit:
Add background, and style dialog boxes and main window

Commit:
Finish Level-10

Commit:
Update README

Commit:
Update code to match coding style

Commit:
Add screenshot

Commit:
Update README.md

Commit:
Add UwuCommand & OwoCommand

Commit:
Add HelloCommand

Commit:
Update README.md

Commit:
Add voice-over

Commit:
Add voice-over

The chatbot is currently silent except for a generic notification sound.

Having a voice will give the chatbot more personality and make it seem
friendlier.

Let's add voice-over for every command and error the bot has.

Commit:
Add assertions

Commit:
Merge pull request #1 from TypeDefinition/branch-A-Assertions

Merge branch-A-Assertions

Commit:
Improve code quality

There are certain try-catch with empty catch blocks.

An explanation comment helps to explain why the catch blocks are empty,
to prevent confusion for future developers.

Lets add an explanation comment to all empty catch blocks.

Commit:
Merge pull request #2 from TypeDefinition/branch-A-CodeQuality

Improve code quality

Commit:
Add help command

Commit:
Merge pull request #3 from TypeDefinition/branch-C-Help

Add help command

Commit:
Add CI

Commit:
Create gradle.yml

Commit:
Remove wrongly added .yml file.

Commit:
Remove wrong CI file.

Commit:
Create gradle.yml

Commit:
Update .yml file

Commit:
Update .yml file

I know this is the least descriptive commit message on the planet but I
need to push in order for GitHub actions to run. The damn build is
failing and I have no idea what the heck I'm doing at this point.

Commit:
Add gradle-wrapper-validation.yml

Commit:
Add CI and gradle files

CI requires gradle files to run.

The gradle files was gitignored, causing the CI build to fail.

Lets unignore gradle files.

Commit:
Fix code style issues

Commit:
Fix unit test code style issues

Commit:
Fix code style issues

Commit:
Fix JavaFX version error

Commit:
Set theme jekyll-theme-tactile

Commit:
Tweak GUI elements to make the app more aesthetically pleasing

Commit:
Upload user guide

Commit:
Add user guide

Commit:
Changed markdown format to GFM so that it renders correctly on GitHub pages

Commit:
Edit user guide title and description

Commit:
Set theme jekyll-theme-time-machine

Commit:
Set theme jekyll-theme-tactile

Commit:
Set theme jekyll-theme-time-machine

Commit:
Add download button to GitHub page

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
update images

Commit:
update
